Added min/max metadata to numerics and a default binning strategy

senior opened the pull request:
This commit adds new min/max metadata to all numeric columns. That
metadata is used in the automatic binning of breakout fields. This
commit also extends MBQL to indicate that a breakout column (or
columns) should be binned using a "default" strategy. The query must
also include the maximum number of bins. Using the number of bins with
the min/max values, the bin width is computed and used in the SQL
query to group data.

-------------------------------------------

This approach to automatic binning is probably the easiest but it does have some drawbacks. The query includes a "max number of bins". If there are no values for a bin, it does not emit a row. This is a by-product of how SQL works. It seems like this might be unintuitive for users. If the user said they wanted 20 bins, my guess is they'd probably like to see the begin/end range of the bin and 0 for the number of entries for the given bin. The breakout_test changes included in this PR show that unintuitive behavior. The test asks for 20 bins, but the data is very clumpy, so all of the data fits in 5 bins. Creating 20 bins with 0s in the empty bins would be a significantly more complicated query for the query processor, but it should be possible.

Something that @salsakran and I have discussed are various ways that users would want to bin discrete values (number of bins, user defined bins, regex on text) in addition to various defaults or predefined binning strategy values. This approach here assumes that the query includes the information needed to execute the query. With this approach, we could still have an admin setup and name an "age range bin" or "version regex" however the querying for that default/predefined bin would happen earlier in the query pipeline. That strategy would be expanded earlier in the pipeline and query processor would still be able to translate the query and execute it straight away. The query would look something like:

```
{:breakout [[:binning-strategy :version :user "version prefix regex"]]}
```

Which would expand internally to:

```
{:breakout [[:binning-strategy :version :regex "^(v.*)/.*" 1]]}
```

We wouldn't need to expose the flexibility to the user via the UI, it could be just an internal detail which allows a clear separation between query construction and user defined settings. It does seem like allowing users to supply this might be a nice future feature though. The user is clicking into a breakout, uses the admin defined regex, they want something a little bit different. The back-end would just get it's regex from the user rather than a setting.

-------------------------------------------
TODO

- [x] Add dimension options to the query_metadata response
- [x] Add dimension info (bounds, strategy used, number of bins etc) to columns in the query response
- [x] Derive min/max from query criteria if present
- [x] Binning should not be allowed when no min/max is present
- [x] If a user is able to issue a binning query without min/max, return a better error message
- [x] `field.default_dimension_option` not implemented
- [x] Handle multiple/conflicting criteria on binned columns
- [x] Support for H2
- [x] Support for MySQL (App DB)
- [x] Support for MySQL (Warehouse DB)
- [x] Support for PostgreSQL (App DB)
- [x] Support for PostgreSQL (Warehouse DB)
- [x] Support for Oracle
- [x] Support for SQL Server
- [x] Support for sqlite
- [x] Support for Vertica
- [x] Support for Presto
- [x] Support for Redshift
- [x] Support for BigQuery
- [x] Do rounded bins work?
- [x] Do we bin by default?
- [x] Bug: ranges starting with 0 are blank
- [x] Bug: axis are off; should we just use ordinal axis and fill missing bins?
- [x] Bug: updating existing filters in QB bar does not work (new value is not applied)
- [x] Add option not to bin
- [x] Improve binning strategy labels
- [x] Disable Heat Maps option
- [x] Won't Fix: after a filter gets removed, the binning isn't reseted correctly to an automatic binning
- [x] Bug: Drill-through in row charts produces an invalid query
- [x] Put breakout granularity in query builder bar, like how we do for time breakouts
- [ ] Fix merge conflicts with the analysis stuff
- [ ] Drill-through on row charts works, but changes the visualization to a normal Bar chart
- [ ] The un-binned state of the breakout label in the QB bar incorrectly shows the FK linkage, but it doesn't do this when the field _is_ binned
- [x] You can't add breakouts to a saved question. When you try to open the field picker, you get the following error:
```
Dimension.js?9b86:122 Uncaught TypeError: Cannot read property 'map' of undefined
    at FieldIDDimension.dimensions (http://localhost:8080/app/dist/app-main.hot.bundle.js?285a42a60f194e62174d:99029:89)
```

**Out of scope for polish committee**
- [ ] Left and right of histograms are showing unnecessary padding
- [ ] Table visualization should show rows for bins with null/zero results if I force a specific number of bins (e.g. 50 bins should show me 50 table rows)
- [ ] Lat/long bins shouldn't have trailing zeros
- [ ] Changing the logic around default binning (right now it bins by default)
- [ ] Heat maps
- [ ] Support for Druid
- [ ] Support for Mongo

**Bugs**
- [x] Shouldn't show binning options for fk/pk types?

iethree committed:
Added min/max metadata to numerics and a default binning strategy

This commit adds new min/max metadata to all numeric columns. That
metadata is used in the automatic binning of breakout fields. This
commit also extends MBQL to indicate that a breakout column (or
columns) should be binned using a "default" strategy. The query must
also include the maximum number of bins. Using the number of bins with
the min/max values, the bin width is computed and used in the SQL
query to group data.

iethree committed:
Number of bins now optional

This adds a new global setting to change the default (currently
8). When the number of bins isn't provided in the query, the default
it used.

iethree committed:
Fix breakout test failure on MySQL

The issue only shows up in Circle CI and appears to be related to
MySQL there returning java.lang.Float values instead of
doubles. Converting these to doubles (via round-to-decimals) before
comparing fixes the issue.

tlrobinson committed:
Merge branch 'multi-metrics-proto' of github.com:metabase/metabase into binning-strategies

tlrobinson committed:
Dimension improvements

tlrobinson committed:
Mock dimension_options until API returns them

iethree committed:
Force the reload of the driver namespaces after plugins load

The compilation order can change from machine to machine. Depending on
when the driver namespaces are loaded relative to when the plugins are
loaded, it can assume drivers are not present even though code is
available via the plugins mechanism.

This commit always reloads the driver namespace after the plugins are
loaded.

iethree committed:
Added dimension_options to the query_metadata response

iethree committed:
Merge remote-tracking branch 'upstream/master' into binning-strategies

tlrobinson committed:
Merge branch 'binning-strategies' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Merge branch 'multi-metrics-proto' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Switch to using dimension_options provided by backend

iethree committed:
Switch to a case based binning, added binning metadata to results

iethree committed:
Switching binning metadata to use underscores

iethree committed:
Sort column dimension_options numerically

iethree committed:
Added min/max values to binning_info

iethree committed:
First cut at constraing min/max from query criteria

senior commented:
Reopening this using binning-strategy branch on the central repo. Old PR is https://github.com/metabase/metabase/pull/5083

iethree committed:
Fields without min/max data should not include dimension options

iethree committed:
Docstring fixups

iethree committed:
Throw a better exception when binning with no min/max

iethree committed:
Added support for multiple min/max affecting criteria clauses

iethree committed:
Switch min/max columns from floats to doubles

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Merge branch 'binning-strategies' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Binned drill through

tlrobinson committed:
Merge branch 'multi-metrics-proto' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Fix graph fill values setting

tlrobinson committed:
Histogram x-axis option

tlrobinson committed:
Fix popover + tooltiptransitions

iethree committed:
Add the "bin-width" binning strategy

In MBQL you can now specify a bin-width binning strategy with a
specified bin-width and the BinnedField will bin using that
information.

This commit also changes the existing behavior of the "default"
binning strategy. When default is specified (for now) this uses the
"num-bins" strategy with the globally set default number of
bins (specified via setting). Specifying num-bins or bin-width
requires an extra parameter specifying the number of bins or the bin
width. Only default uses the global setting.

tlrobinson committed:
Heat maps v2

tlrobinson committed:
Merge branch 'binning-strategies' of github.com:metabase/metabase into heat-maps-2

iethree committed:
Switched bin calculation from case to equation

This commit also changes from an aliased projection, which does not
work on all SQL databases to a repeated expression. The equation now
appears in the SELECT, GROUP BY and ORDER BY. This whould work across
SQL databases.

iethree committed:
Add latitude/longitude specific binning options

iethree committed:
Add num-bins test

iethree committed:
WIP - Add type based defaults for coordinates

tlrobinson committed:
Merge branch 'binning-strategies' of github.com:metabase/metabase into heat-maps-2

tlrobinson committed:
Merge remote-tracking branch 'origin/multi-metrics-proto' into binning-strategies-frontend

tlrobinson committed:
Refactor drill downs, implement for heat maps, misc heatmap improvements

tlrobinson committed:
Merge branch 'binning-strategies-frontend' into heat-maps-2

tlrobinson committed:
Merge branch 'binning-strategies' of github.com:metabase/metabase into binning-strategies-frontend

iethree committed:
Revert "Force the reload of the driver namespaces after plugins load"

This reverts commit 8f552032895121b851e9f37aed7dacb70ee7a64c.

iethree committed:
Merge remote-tracking branch 'upstream/master' into binning-strategies

iethree committed:
Feature flag binning

iethree committed:
Code cleanup to prep for review [ci drivers]

iethree committed:
Fix up oracle tests

iethree committed:
Add min/max to presto expected test results [ci drivers]

iethree committed:
Remove min/max from mongo's expected field values [ci drivers]

iethree committed:
Merge remote-tracking branch 'upstream/master' into binning-strategies

iethree committed:
Add binning support for presto and bigquery [ci drivers]

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into binning-strategies

tlrobinson committed:
Merge branch 'binning-strategies' into binning-strategies-frontend

tlrobinson committed:
Merge branch 'binning-strategies-frontend' into heat-maps-2

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Merge branch 'binning-strategies-frontend' of github.com:metabase/metabase into heat-maps-2

tlrobinson committed:
Clean up drilldown.js, add spec, misc other improvements

tlrobinson committed:
Fix fieldOptions not including fk fields

tlrobinson committed:
Fix issue with dashcard resize widget and leaflet map resizing

tlrobinson committed:
Fix issue with loading a Leaflet heatmaps a second time

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into binning-strategies

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Merge branch 'binning-strategies' of github.com:metabase/metabase into binning-strategies-frontend

tlrobinson committed:
Merge branch 'binning-strategies-frontend' of github.com:metabase/metabase into heat-maps-2

iethree committed:
Return binning strategy in the field's binning_info

This hadn't been updated to allow for the two binning strategies that
we now support.

iethree committed:
Add default_dimension_option for binning/dates

This commit adds a new keypair to the fields map called
"default_dimension_option". For datetimes, this is grouping by day,
for binning this is the "default" binning strategy.

sbelak committed:
Added some simple heuristics for nicer bin sizes & bounds

sbelak committed:
Merge branch 'binning-strategies' of github.com:metabase/metabase into binning-strategies

sbelak commented:
Added some simple heuristics to make bin sizes and bounds more humane. The general idea is:

1) if we have fixed width round both ends to the same resolution (ie. for a width of 20 round to 20)
2) if we have fixed number of bins, start with the smallest possible bin width (ie. max-min/N), enlarge it to the first "pleasing" number (say 5, 10, 20, 50, 100, ...), go to 1)

iethree committed:
Fixup breakout test failures

iethree committed:
Merge remote-tracking branch 'upstream/master' into binning-strategies [ci drivers]

iethree committed:
Minor cleanup/refactoring for binning

iethree committed:
Added a missing require for an interface defrecord

iethree committed:
Merge remote-tracking branch 'upstream/master' into binning-strategies

attekei committed:
Merge branch 'binning-strategies' into heat-maps-2

attekei committed:
Fix linter errors, fix Jest tests

attekei committed:
Don't return binned fields from FieldList without subdim selector

attekei committed:
Fix lat/lon formatting Karma test [ci e2e]

attekei committed:
Support `transformResponse` request option in integrated tests [ci e2e]

iethree committed:
Add a "Don't bin" strategy to the table API response

The mbql clause is nil, so when chosen it will not change or add to
the query in any way.

mazameli committed:
change strings, remove dead code

mazameli committed:
merge

attekei committed:
Fix an unresolved merge conflict

attekei committed:
Put 'Don't bin' as the last option

attekei committed:
Fix binning option names in frontend

attekei committed:
Show 'Unbinned' in hover, show FK name correctly in breakout widget

attekei committed:
Don't show FK source for date/binning breakouts

attekei committed:
Fix issue with being not able to update filters

attekei committed:
Fix issue with not being able to add breakouts to nested queries

tlrobinson committed:
Fix binned row chart tooltip/drillthrough

attekei committed:
Minor optimization to row chart drill-through/tooltip logic

attekei committed:
Disable heat and grid map visualization types

attekei committed:
Fix binning drill-through

attekei committed:
Make test db fixture contain complete db analysis results

attekei committed:
Fix drilldown test, convert Dimension test to integrated test

attekei committed:
Add integration test for input filters

attekei committed:
Add tests for adding a binned breakout in QB

sbelak committed:
Fix a bug in order-of-magnitude (x<=0) and moved it to util

attekei committed:
Add test for location breakout with binning

attekei committed:
Fix lint and Flow errors

attekei committed:
Merge branch 'binning-strategies-frontend-fixes' into binning-strategies

attekei committed:
Run integrated tests [ci e2e]

attekei committed:
Add tests for drill-through, format the binned coordinates [ci e2e]

iethree committed:
Merge remote-tracking branch 'upstream/master' into binning-strategies

iethree committed:
Switch from min/max value on field to fingerprints

This commit reworks binning to use the new fingerprint field added
during analysis, rather than the (no longer existing) min/max columns
on the metabase_fields table.

sbelak committed:
Fix failing test (due to double -> int change) [ci all]

iethree committed:
Round fingerprints (in test) to 3 decimal places [ci drivers]

Oracle specifically has long unrounded decimals that cause tests to
fail. This munges results to round the decimals to 3 places before comparing.

iethree committed:
Merge remote-tracking branch 'upstream/master' into binning-strategies

iethree committed:
Trigger ci [ci all]

iethree committed:
Merge remote-tracking branch 'upstream/master' into binning-strategies
